File: clojure-solutions/src/year-2015/day6.clj
(ns advent-of-code.2015.day6)


(def input (->> (slurp "resources/2015/day6.txt")
                (clojure.string/split-lines)
                (map #(map read-string (clojure.string/split % #"\s+")))))

(defn produce-indices [x1 x2 y1 y2]
  (for [i (range x1 (inc y1)) j (range x2 (inc y2))]
    [i j]))

(defn turn-off [x]
  (cond
    (nil? x) 0
    (<= x 0) 0
    :else (dec x)))

(defn turn-on [x]
  (cond
    (nil? x) 1
    :else (inc x)))

(defn toggle [x]
  (cond
    (nil? x) 2
    :else (+ x 2)))

(defn update-grid [g k instruction]
  (case instruction
    :on (update g k turn-on)
    :off (update g k turn-off)
    (update g k toggle )))

(defn process [grid [inst a b c d]]
  (let [xs (produce-indices a b c d)]
    (reduce (fn [acc k] (update-grid acc k inst)) grid xs)))

(defn solve []
  (->> (reduce process {} input)
       (vals)
       (reduce +)))

File: python/2015/day6.py
import re
from collections import defaultdict


with open("./day6.txt", "r") as infile:
    data = infile.read().splitlines()
    d = defaultdict(int)

    for line in data:
        xs = re.findall(r"\d+", line)
        x1, y1, x2, y2 = [int(x) for x in xs]
        for i in range(x1, x2 + 1):
            for j in range(y1, y2 + 1):
                if "turn on" in line:
                    d[(i, j)] += 1
                elif "turn off" in line:
                    d[(i, j)] = 0 if d[(i, j)] <= 1 else d[(i, j)] - 1
                elif "toggle" in line:
                    d[(i, j)] += 2


    total = 0
    for v in d.values():
        total += v
    print(total)
                    
